Introduce StableHLO interpreter API

penagos opened the pull request:
Today, it isn't practical to call the StableHLO reference interpreter. Invoking the interpreter requires going through the MLIR translate infrastructure (i.e. via `interpretRegistration` in `StablehloTranslateMain.cpp`), and does not currently support specifying user inputs or doing any meaningful post processing on the model outputs short of printing them to screen. This makes it difficult to build tooling which uses the reference interpreter.

This PR proposes a new reference interpreter API which facilitates the running the StableHLO reference interpreter through a simple, unified interface. At a high level, the proposed interface allows clients to specify a set of inputs (in `InterpreterValue` format), and an `InterpreterConfiguration` object (see below). The proposed interface invokes the reference interpreter and returns a set of outputs (in `InterpreterValue` format). The new interface would live in `stablehlo/reference/InterpreterApi.h` and would expose the following interface:

```
/// Invoke the StableHLO reference interpreter with the given parsed MLIR
/// module input and provided inputs. Returns a list of interpreter outputs.
/// Can optionally pass a fallback interpreter callback which executes when no
/// builtin kernels are matched.
llvm::ErrorOr<SmallVector<InterpreterValue>> evalModule(
    ModuleOp module, ArrayRef<InterpreterValue> inputs,
    const InterpreterConfiguration &config);
```


### Interpreter configuration
As part of this PR, there is an `InterpreterConfiguration` struct which can be used to configure the StableHLO interpreter. Encapsulating configuration state into a singular object allows the interface to remain stable as new configuration options are added. Initially, the configuration object will contain a handful of options:

```
struct InterpreterConfiguration {
  InterpreterConfiguration()
      : fallback(std::make_unique<InterpreterFallback>()) {}

  /// If specified, the directory to which StableHLO interpreter tensors will
  /// be serialized to disk.
  std::string probeInstrumentationDir = "";

  /// Use the specified named function as the main entrypoint into a module.
  /// Defaults to `main` for modules with multiple functions. If a module only
  /// contains 1 function and the default `main` value is used, the singular
  /// function will be used as the entrypoint (irrespective of a function name
  /// match).
  std::string mainFunction = "main";

  /// If specified, use the callback to run on ops which do not have a
  /// registered kernel.
  std::unique_ptr<InterpreterFallback> fallback;
};
```

Note that all of these configuration values are optional, and it is possible to run the interpreter with a default initialized `InterpreterConfiguration` struct (which will by default, not provide a fallback callback).

### Interpreter fallback callback
When defining an interpreter fallback callback, the `InterpreterFallback` struct can be extended:

```
class InterpreterFallback {
 public:
  /// Custom op kernels for any user specified ops not found in the StableHLO
  /// op dialect or StableHLO interpreter dialect.
  virtual llvm::Error operator()(Operation &op, Scope &scope, Process *process);

  virtual ~InterpreterFallback() = default;
};
```
Note that the user specified fallback need not contain any `InterpreterOp` kernels, these will always run before the user specified callback.

Currently, the proposed change exposes `InterpreterConfiguration` and the current `mlir::func::FuncOp` being evaluated to the fallback interface to support the existing flow. Consequently, there’s a bit of gymnastics in `stablehlo::runInterpreter` which I think could be improved on. I'm open to ideas on how we could better handle this.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,215 @@
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/reference/NumPy.h"
+#include "stablehlo/reference/Ops.h"
+#include "stablehlo/tests/CheckOps.h"
```
I don't think CheckOps should ship with the default implementation.

I believe the default fallback should just be "InterpreterOps", and should allow subclassing to support InterpreterOps + their own dialect ops / custom call ops.

Could have methods for:
- `handleInterpreterOp` (this should be marked final, not overridable)
- `handleUnregisteredOp` (check dialect could be added here)
- `handleCustomCallOp` (any additional handling)

Or could be a map of `op_name -> fallbackHandler`, then users could add their own `addCheckDialectFallbacks(handler)` methods to populate the fallback handler class.

But for now I think `CheckOps` should remain a feature of `StablehloTranslateMain`, as exposing them to the wild has broader maintenance implications.

penagos replied:
Hm, how about just implementing the `InterpreterOp` in the default fallback functor in lieu of 3 separate callbacks? And then leaving the `CheckOp` kernels within the impl of `StablehloTranslateMain`?

GleasonK commented on stablehlo/tests/CheckOps.cpp:
```diff
@@ -128,6 +136,29 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensorOrError);
 }
 
+llvm::Error evalCustomCallCheckEq(stablehlo::CustomCallOp op,
```
Similarly, let's keep this custom call handling in `StablehloTranslateMain` as it was. The custom call is a bit of a hack for `testdata` testing, it shouldn't be used more broadly than our tests. A better solution would be adding a pass that converts custom_call -> check dialect, but no need to do it now

GleasonK reviewed (changes requested):
Proposed refactoring:

```
InterpreterApi.h/cpp (_only_ handles StableHLO, no check or custom call)
StablehloTranslateMain.cpp (implements InterpreterFallback with Check and custom call)
```

If we need to expose CheckDialect in the future, we can, but for now lets consider that an implementation detail of StableHLO test infrastructure. Non-test code depending on test code is a bit of an antipattern

ghpvnist commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,96 @@
+  void setConfig(const InterpreterConfiguration &cfg) { config = &cfg; }
+
+  /// Set the topmost currently executing function in the module.
+  void setFcn(func::FuncOp fcn) { currentFcn = fcn; }
```
Proposal:
```
void setFunction(func::FuncOp funcOp) { currentFunction = funcOp; }
```

Abbreviating this does not give us much mileage, plus the code in the implementation
```
config.fallback->setFcn(funcOp);
```
would read nicely:
```
config.fallback->setFunction(funcOp);
```

ghpvnist commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,96 @@
+  virtual ~InterpreterFallback() = default;
+
+  /// Set the user provided interpreter configuration.
+  void setConfig(const InterpreterConfiguration &cfg) { config = &cfg; }
```
`cfg` -> `config` to be consistent with name defined in struct and arg name used in `runInterpreter`

ghpvnist commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,96 @@
+  llvm::StringMap<int32_t> instrumentedTensors;
+};
+
+struct InterpreterConfiguration {
```
Let's alphabetically order them for ease of maintenance.

ghpvnist commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,166 @@
+                                    debugString(op).c_str());
+}
+
+llvm::ErrorOr<SmallVector<InterpreterValue>> runInterpreter(
```
```
llvm::ErrorOr<SmallVector<InterpreterValue>> runInterpreter(
    const std::string &mlir, ArrayRef<InterpreterValue> inputs,
    const InterpreterConfiguration &config) {
  llvm::SourceMgr source_mgr;
  source_mgr.AddNewSourceBuffer(llvm::MemoryBuffer::getMemBuffer(mlir),
                                llvm::SMLoc());

  MLIRContext context;
  OwningOpRef<ModuleOp> module(parseSourceFile<ModuleOp>(source_mgr, &context));

  return runInterpreter(module.get(), inputs, config);
}
```

ghpvnist commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,166 @@
+  return handleOp(op, process, scope);
+}
+
+llvm::Error InterpreterFallback::handleOp(Operation &op, Process *process,
```
I don't think we need this abstracted out yet. Since we don't handle them in multiple ways, let's keep the error message inlined?

penagos replied:
The initial idea behind splitting this was that if/when user specify a custom fallback, they need not worry about explicitly inheriting or calling the interpreter ops fallback. That is, interpreter ops would always have a kernel defined.

Were you thinking of having the interpreter op kernels be optional in a singular overridable callback?

ghpvnist commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,96 @@
+/// Base interpreter fallback callback functor to run when no registered kernels
+/// are found for a given StableHLO operation.
+struct InterpreterConfiguration;
+struct InterpreterFallback {
```
Let's convert this to a class per https://stackoverflow.com/questions/54585/when-should-you-use-a-class-vs-a-struct-in-c

ghpvnist reviewed (changes requested):
Went through first pass. I need a bit more time to think about the `InterpreterFallback` API.

penagos commented on stablehlo/tools/StablehloTranslateMain.cpp:
```diff
@@ -15,29 +15,21 @@ limitations under the License.
 #include "stablehlo/dialect/StablehloOps.h"
-#include "stablehlo/dialect/VhloOps.h"
-#include "stablehlo/reference/Errors.h"
+#include "stablehlo/dialect/Version.h"
```
This include is needed by:
```
 targetVersion = vhlo::Version::getCurrentVersion().toString();
```

ghpvnist replied:
Ah okay, I didn't see this being used in the diffs initially.

ghpvnist commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,99 @@
+  func::FuncOp currentFunction;
+
+ private:
+  /// Counts how many times a given probe_id has been used while profiling.
```
Could you elaborate on the comment from the viewpoint of the API; from referencing this doc only, I don't know what `probe_id` is, which is an implementation detail of the ProbeOp.

ghpvnist reviewed (changes requested):
I reviewed the rest and left a few comments. PTAL.

penagos commented on BUILD.bazel:
```diff
@@ -392,6 +395,32 @@ cc_library(
+        ":register",
+        "@llvm-project//llvm:Support",
+        "@llvm-project//mlir:FuncDialect",
+        "@llvm-project//mlir:IR",
```
Removed `:register` dependency, but I do think we need the others due to existing `#include` in the impl file.

penagos commented on BUILD.bazel:
```diff
@@ -392,6 +395,32 @@ cc_library(
+        ":reference_process",
+        ":reference_scope",
+        ":reference_tensor",
+        ":register",
```
I think we still need some of the `reference_*` dependencies for the header includes.

GleasonK commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,95 @@
+  virtual llvm::Error handleOp(Operation &op, Process *process, Scope &scope);
+
+  /// The user provided interpreter configuration.
+  const InterpreterConfiguration *config;
```
I don't like this setup where we have an `InterpreterConfig` and an `InterpreterFallback` which both must point at eachother if a fallback exists.

Can we make this a unidirectional dependency where a Config has a Fallback and that's it. If there's ever a reason for the back-pointer, I think we should modify other APIs to accept a config instead of a fallback.

If we do this refactoring we can:
- Remove the forward declaration. 
- Update InterpreterConfig to take an InterpreterFallback by reference (open to thoughts on this, I'm ok with pointer, just worth noting that under the current setup we throw away a fallback every time we use `StablehloTranslate` since we override it)

penagos replied:
I agree, it's a bit confusing of a design. I've refactored it to pass through a config instead to `eval` as the default fallback does need some configuration state (so the fallback functor will still take a config, but it doesn't have a chance of being "outdated" by the toplevel caller not correctly calling `setConfig` before the fallback runs).

However, I think there's an opportunity to further clean up some of the current interpreter design here. When changing the interface, I noticed there's a multiple op kernels which take a `Process` and `Scope`. For example, `evalAllReduceOp`, `evalCaseOp`, `evalIfOp` to name a few, which later go on to recursively call back into teh top level `eval`. However, they don't preserve the current top level fallback function.

For example, `evalIfOp` calls back into `eval` for both true/false branches:

```
  return pred.get({}).getBooleanValue()
             ? eval(trueBranch, {}, /*config=*/nullptr, process, &scope)
             : eval(falseBranch, {}, /*config=*/nullptr, process, &scope);
```

But presumably, if either branch had an op which was handled in the top level specified fallback, that wouldn't work. I'm not sure if this is by design, or an omission. If the latter, since `eval` would also need to take an `InterpreterConfiguration` for properly handling interpreter instrumentation, would it be worthwhile to instead encapsulate `Process`, `Scope` and `InterpreterConfiguration` into something like an `InterpreterState` struct?

GleasonK replied:
Ah ignore my other comment, I didn't realize these bugs existed prior to your change. This is on us to clean up. I'll make a ticket with this refactoring / bug fixing

GleasonK replied:
Created https://github.com/openxla/stablehlo/issues/1820

Thanks for calling this out. Feel free to resolve comment, I'm currently impartial to the refactoring you've mentioned. If we end up needing to refactor all callsites with a new signature for this context we may reconsider this

GleasonK commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  /// Set the topmost currently executing function in the module.
+  void setFunction(func::FuncOp function) { currentFunction = function; }
```
Let's remove the `currentFunction` field and this setter. I believe this value will currently be wrong any time we call a function..since it is only updated at the top level?

Instead we can use `op.getParentOfType<func::FuncOp>()` to get the nearest enclosing FuncOp from the IR when it is needed.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,126 @@
+    if (numFuncs > 1 && funcOp.getSymName() != config.mainFunction)
+      return WalkResult::advance();
+
+    if (config.fallback) config.fallback->setFunction(funcOp);
```
Delete per other feedback.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,126 @@
+    results = stablehlo::eval(funcOp.getBody(), inputs, /*process=*/nullptr,
+                              /*parent=*/nullptr, *config.fallback);
+
+    if (config.stream)
```
Let's make this even more flexible, and let `config` take a `resultHandler` function which takes results and returns a `llvm::Error`. The `StablehloTranslate` impl of this function can call print and return success.

This method will just throw the error if there's a failure in the callback.

penagos replied:
Thinking about this more, we probably don't need this complexity at all. I.e. we can move this specific use case into the `interpretRegistration` callback in `StablehloTranslateMain.cpp`.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,126 @@
+  SmallVector<InterpreterValue> results;
+  config.fallback->setConfig(config);
+
+  auto walkResult = module.walk([&](func::FuncOp funcOp) {
```
Per above comment, I don't think we need this walk, if the goal is to just find the main function for executing, lets get it above and just call eval here.

GleasonK commented on stablehlo/reference/InterpreterApi.h:
```diff
@@ -0,0 +1,95 @@
+/// module input and provided inputs. Returns a list of interpreter outputs.
+/// Can optionally pass a fallback interpreter callback which executes when no
+/// builtin kernels are matched.
+llvm::ErrorOr<SmallVector<InterpreterValue>> runInterpreter(
```
I've been trying to think of a better name, and so far I've come up with:
- referenceEval
- evalModule
- runModule
- interpret
- interpretModule

I think `evalModule` is my preference since it is more clear what the inputs are and what it is doing.

penagos commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,126 @@
+  bool hasMain = false;
+
+  assert(!config.mainFunction.empty() && "Main function name cannot be empty");
+  module.walk([&](func::FuncOp funcOp) {
```
I think to preserve the current behavior (i.e. for existing LIT tests), the semantics of the function need to be if there is only 1 function in the module, use that as the main function. Otherwise, use the provided `mainFunction` string to break ties when there are more than 1 function. Slightly tweaked proposed change with this.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -0,0 +1,114 @@
+func::FuncOp getMainFunction(ModuleOp module, StringRef mainName) {
+  auto functions = module.getOps<func::FuncOp>();
+
+  // If the module has 1 function only, use it as the main function.
+  if (std::distance(functions.begin(), functions.end()) == 1) {
+    return *functions.begin();
+  }
```
Nit - there is something odd that this function takes a `mainName` that is ignored if there is only 1 function.

```
func.func @not_main() { ... }
```

If I call `evalModule` for `main` on the above module.. it still runs.

GleasonK replied:
Maybe we add a check that the user didn't specify a custom main function. I.e. only use the 1 function if the function we're looking for is `main`.

Let's move this check below the iteration over all ops, and add a check for:

```cpp
  for (auto funcOp : functions)
    if (funcOp.getSymName().equals(mainName)) return funcOp;
 
bool isSingleFunction = std::distance(functions.begin(), functions.end()) == 1;
bool isDefaultLookup = mainName == "main";
if (isSingleFunction && isDefaultLookup)  return *functions.begin();
```

GleasonK commented on stablehlo/reference/Ops.cpp:
```diff
@@ -817,18 +818,18 @@ SmallVector<InterpreterValue> eval(
       auto &cond = whileOp.getCond();
       auto &body = whileOp.getBody();
-      auto results = evalWhileOp(operand, cond, body, process, scope, fallback);
+      auto results = evalWhileOp(operand, cond, body, process, scope, config);
```
Nit, lets be uniform with where the config arg appears. In `eval` it is before process/scope, in `evalWhileOp` it is a trailing argument. I don't have a strong preference on which position it should be, just that it is consistent.

penagos replied:
Decided to keep it before `Process` as I think once #1820 is addressed, we can make it a `const InterpreterConfiguration&` and keep it before any other optional pointer parameters.

GleasonK commented on stablehlo/reference/Ops.cpp:
```diff
@@ -199,10 +200,10 @@ SmallVector<Tensor> split(const Tensor &x, int64_t numResults, Axis axis,
-    llvm::function_ref<llvm::Error(Operation &, Process *, Scope &)> fallback) {
+SmallVector<InterpreterValue> eval(Region &region,
+                                   ArrayRef<InterpreterValue> args,
+                                   const InterpreterConfiguration *config,
```
Why not make this a reference? It looks like the important part is that Fallback is optional. I think making config optional and fallback optional makes the code more complicated.

```cpp
if (config && !config->fallback)
==>
if (!config->fallback)

foo(...,/*config=*/nullptr)
==>
foo(..., /*config=*/{})
```

GleasonK replied:
Also for the places where config is nullptr.. is this the proper behavior? I feel like this should be passed through regions? I.e. probe in the body of a `stablehlo.map` operation looks like it wouldn't work currently

penagos replied:
Yup, https://github.com/openxla/stablehlo/pull/1797#discussion_r1378996442 is the reason this is an optional pointer for the time being, since otherwise the other interfaces that call into `eval` today would need to be updated to take a `config` as described in #1820. I think maybe this could be changed to required pass by reference once those ops are updated to take the fallback interface correctly.

GleasonK commented on stablehlo/reference/Ops.h:
```diff
@@ -206,10 +206,9 @@ Tensor evalXorOp(const Tensor &lhs, const Tensor &rhs, ShapedType resultType);
-    function_ref<llvm::Error(Operation &, Process *, Scope &)> fallback =
-        nullptr);
+    Region &region, ArrayRef<InterpreterValue> args,
+    const InterpreterConfiguration *config = nullptr,
```
Per earlier comment, I don't think this should be optional here. Let's flip to reference and pass the context through.

GleasonK reviewed (approved):
LGTM. Thanks for the contribution!

ghpvnist reviewed (approved):
Looks like we need to add InterpreterApi dependency. All looks good to me as well otherwise. Thanks again for the contribution!

penagos commented:
Refactored the interface to avoid the build issue and to try and make it a bit cleaner. Instead of having 2 functions in the fallback callback interface functor, there's just the singular `operator()` now, and I've reverted the `evalXX` changes to pass around a config, and instead pass around a fallback callback pointer as before.

Now, the interpreter interface will create a lightweight wrapper that sits ontop of the optionally overridable fallback callback, to handle kernels for probe/parallel ops. This should avoid the circular include dependency, and somewhat simplify the interface.

GleasonK commented on stablehlo/reference/InterpreterApi.cpp:
```diff
@@ -44,47 +43,63 @@ func::FuncOp getMainFunction(ModuleOp module, StringRef mainName) {
+                                "interpreter.run_parallel");
+    }
+
+    return (*config.fallback)(op, scope, process);
```
If config points back to this class, which I think is a typical usage(?) wouldn't this recurse infinitely?

GleasonK replied:
Ah. This is an internal implementation detail. Please add a comment to the class saying "DefaultInterpreterFallback is an implementation detail of run module. It takes in an InterpreterConfiguration which can have user-implemented fallbacks"

As of now, this setup means that a custom op cannot have a region that uses a default callback (i.e. my.op with a region containing a probe op). This is probably an OK limitation for now, but would require refactoring into something more composable if this ever becomes a requirement.
